### PR TITLE
Trampoline to a base method implementation when deleting an override

### PR DESCRIPTION
**Pull request by davidwengier**
Part of #59264

Given:
```
class B
{
    public virtual void M(int x)
    {
    }
}

class C : B
{
    public override void M(int x)
    {
    }
}
```

When deleting `C.M`, instead of emitting `throw new MissingMethodException()` this will now emit `base.M(x);`

### CODE REVIEW

File: src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs [hunk 1]
```diff
@@ -547,8 +547,24 @@ protected override void CreateIndicesForNonTypeMembers(ITypeDefinition typeDef)
...
+                    var overriddenMemberDef = (IMethodReference?)_changes.TryGetBaseImplementationDefinition(methodDef, out var isThisAssembly);
+
+                    uint token = 0;
+                    if (overriddenMemberDef is not null)
+                    {
+                        // Because the deleted symbols, and hence the overridden member symbol, come from a previous
+                        // generation, the normal IL emit process doesn't know how to deal with it, and will always
+                        // think its a new MemberRef row that needs to be added. We have enough information now though,
+                        // to make the correct decisions, so we just emit the raw token for the right method to call
+                        // directly.
```
Comment (davidwengier) [inline src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs:559]: This is the explanation for why this is potentially dodgy, using the token directly.

I've put this up as a draft just to get sign off on the idea, or discuss. Obviously if this approach is okay, I'll clean up the commented out `EmitToken` call in `DeletedMethodBody` :)

File: src/Compilers/Core/Portable/Emit/EditAndContinue/SymbolChanges.cs [hunk 2]
```diff
@@ -423,6 +423,44 @@ SymbolChange fixChangeIfMemberIsReAdded(IDefinition item, SymbolChange change, F
...
+                throw ExceptionUtilities.Unreachable();
+            }
+
+            // We only want to return base definitions that are actually callable
+            if (baseSymbol is null || baseSymbol.IsAbstract)
+            {
+                return null;
+            }
+
+            isThisAssembly = baseSymbol.ContainingAssembly == symbol.ContainingAssembly;
```
Comment (davidwengier) [inline src/Compilers/Core/Portable/Emit/EditAndContinue/SymbolChanges.cs:459]: It was surprising to me, but even when `baseSymbol` is a `PEMethodSymbol`, its still an `IMethodDefinition`, so this was the only way to tell if the base call should be a ref or a def. I was originally planning to check for `IMethodDefinition` vs `IMethodReference`

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 3]
```diff
@@ -14901,6 +14899,136 @@ .maxstack  8
...
+                            }
+                        }
+
+                        class C : B
+                        {
+                            public override void M(int x)
+                            {
+                            }
+                        }
+                        """,
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs:14921]: >""", [](http://example.com/codeflow?start=24&length=4)

We should test with generic type arguments:
```csharp
class B<T>
{
    public virtual void M(T t) { }
}
class C : B<int>
{
    public override void M(int i) { }
}
```

File: src/Compilers/Core/Portable/Emit/EditAndContinue/DeletedMethodBody.cs [hunk 4]
```diff
@@ -68,15 +68,34 @@ public DeletedMethodBody(DeletedMethodDefinition methodDef, EmitContext context)
...
         {
             var missingMethodExceptionStringStringConstructor = context.Module.CommonCompilation.CommonGetWellKnownTypeMember(WellKnownMember.System_MissingMethodException__ctor);
             Debug.Assert(missingMethodExceptionStringStringConstructor is not null);
 
             var builder = new ILBuilder((ITokenDeferral)context.Module, null, OptimizationLevel.Debug, false);
-            builder.EmitOpCode(System.Reflection.Metadata.ILOpCode.Newobj, 4);
-            builder.EmitToken(missingMethodExceptionStringStringConstructor.GetCciAdapter(), context.SyntaxNode!, context.Diagnostics);
-            builder.EmitThrow(isRethrow: false);
+
+            // If we're deleting a method that overrides a base method then we trampoline to it,
```
Comment (cston) [inline src/Compilers/Core/Portable/Emit/EditAndContinue/DeletedMethodBody.cs:78]: >If we're deleting a method that overrides a base method then we trampoline to it, [](http://example.com/codeflow?start=15&length=81)

How important is this scenario? In short, why not just throw a MissingMethodException when deleting an override?

Comment (davidwengier) [reply src/Compilers/Core/Portable/Emit/EditAndContinue/DeletedMethodBody.cs:78]: It just makes things a lot nicer for users. eg, if I had 3 virtual methods in a chain, each one calling `base`, and I remove the middle layer, why should a call to the highest layer throw? Or code that operates on a base class throw if I change something in a child class?

Comment (davidwengier) [reply src/Compilers/Core/Portable/Emit/EditAndContinue/DeletedMethodBody.cs:78]: Another example of where trampolining would be helpful, though not directly solved by this PR, would be https://github.com/dotnet/roslyn/issues/65442, where its not obvious to the user than an edit they made to a method signature, causes a new method to be emitted

**Comment by davidwengier** [comment]
Just cleaning up an old PR

**Pull request closed without merging**